scrapy/utils/datatypes.py

```diff
@@ -63,6 +79,36 @@ def pop(self, key, *args):
         return dict.pop(self, self.normkey(key), *args)
 
 
+class CaseInsensitiveDict(collections.UserDict):
+    """A dict-like structure that accepts strings or bytes as keys and allows case-insensitive lookups.
+
+    It also allows overriding key and value normalization by defining custom `normkey` and `normvalue` methods.
+    """
+
+    def __getitem__(self, key: AnyStr) -> Any:
+        return super().__getitem__(self.normkey(key))
+
+    def __setitem__(self, key: AnyStr, value: Any) -> None:
+        super().__setitem__(self.normkey(key), self.normvalue(value))
+
+    def __delitem__(self, key: AnyStr) -> None:
+        super().__delitem__(self.normkey(key))
+
+    def __contains__(self, key: AnyStr) -> bool:  # type: ignore[override]
```

Comment:
Even if this goes against [Liskov substitution principle](https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides), it makes sense because being case-insensitive only applies to str or bytes, not to all hashable types.

Reply:
I wonder if it's possible to inherit from some type-enabled dict